### PR TITLE
Remove -enable-ossa-modules for Synchronization and Distributed 

### PR DESCRIPTION
**Pull request by meg-gupta**
`Synchronization` and `Distributed` are serialized in OSSA, they import `Darwin` which is non-ossa. With implicit modules builds, this results in a one time recompilation of `Darwin` with `-enable-ossa-modules`. 
However with Explicit Build Modules, all dependencies are treating equally and `-enable-ossa-modules` will be passed on only when present in the original compiler invocation. 

This PR removes `-enable-ossa-modules` for `Synchronization` and `Distributed` and adds a diagnostic whenever we are trying to import a non-ossa module into an ossa module when Explicit Build Modules are enabled. 

As a consequence one of the perf constraint tests in `Synchronization` started failing due to presence of copies which was previously eliminated due to being in ossa. I added @_transparent to the relevant stdlib function - `Atomic.deinit` to workaround this since `@_transparent` functions are always serialized as ossa irrespective of the presence of `-enable-ossa-modules`. 

This is a temporary fix until ossa modules are default. 

Fixes rdar://138407068

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Comment by meg-gupta** [comment]
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/9492

@swift-ci Please test

**Comment by Azoy** [comment]
Is this really easier than just adding `-enable-ossa-modules` to the `Darwin` module? Do you have an estimated timeline of when `-enable-ossa-modules` will be turned on by default?

File: stdlib/public/core/CMakeLists.txt [hunk 1]
```diff
@@ -353,6 +353,9 @@ set(swiftCore_common_options
                   SWIFT_COMPILE_FLAGS
                     ${swift_stdlib_compile_flags} -Xcc -DswiftCore_EXPORTS ${SWIFT_STANDARD_LIBRARY_SWIFT_FLAGS})
 
+# Enable OSSA Modules for swift stdlib
```
Comment (Azoy) [inline stdlib/public/core/CMakeLists.txt:356]: I think you can remove this since it comes for free with the changes in `stdlib/cmake/modules/AddSwiftStdlib.cmake`.

Comment (meg-gupta) [reply stdlib/public/core/CMakeLists.txt]: Done

**Comment by meg-gupta** [comment]
@Azoy Adding `-enable-ossa-modules` to `Darwin` will also need adding it to its dependencies which are multiple in number.

File: lib/Serialization/SerializedModuleLoader.cpp [hunk 2]
```diff
@@ -1190,7 +1190,12 @@ void swift::serialization::diagnoseSerializedASTLoadFailure(
                        moduleBufferID);
     break;
   case serialization::Status::NotInOSSA:
-    // soft reject, silently ignore.
+    // Diagnose only when explicit build modules is enabled
+    if (Ctx.SerializationOpts.ExplicitModuleBuild) {
+      Ctx.Diags.diagnose(diagLoc,
```
Comment (tshortli) [inline lib/Serialization/SerializedModuleLoader.cpp:1195]: Why is this only diagnosed for explicit module builds?

File: include/swift/Serialization/SerializationOptions.h [hunk 3]
```diff
@@ -24,144 +24,144 @@
 
 namespace swift {
 
-  class SerializationOptions {
-    SerializationOptions(const SerializationOptions &) = delete;
-    void operator=(const SerializationOptions &) = delete;
+class SerializationOptions {
```
Comment (tshortli) [inline include/swift/Serialization/SerializationOptions.h:27]: Nit: I assume clang-format fixed the formatting here for the whole class. It probably needed to happen eventually but it would be nice for it all to be fixed in a separate commit so that it's clearer what's happening in the PR

**Comment by meg-gupta** [comment]
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/9492

@swift-ci Please test

File: lib/Frontend/ModuleInterfaceLoader.cpp [hunk 4]
```diff
@@ -1785,8 +1785,11 @@ void InterfaceSubContextDelegateImpl::inheritOptionsForBuildingInterface(
   GenericArgs.push_back("-disable-objc-attr-requires-foundation-module");
 
   // If we are supposed to use RequireOSSAModules, do so.
-  genericSubInvocation.getSILOptions().EnableOSSAModules =
-      bool(RequireOSSAModules);
+  if (RequireOSSAModules) {
+    genericSubInvocation.getSILOptions().EnableOSSAModules = true;
```
Comment (tshortli) [inline lib/Frontend/ModuleInterfaceLoader.cpp:1789]: Does this have a test?

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/Serialization/SerializedModuleLoader.cpp:1195]: I have left the behavior as is for implicit module builds. It's useful for testing `-enable-ossa-modules` on a module level.

File: include/swift/AST/DiagnosticsSema.def [hunk 5]
```diff
@@ -901,6 +901,9 @@ ERROR(serialization_target_too_new_repl,none,
       "deployment target of %0 %3: %4",
       (StringRef, llvm::VersionTuple, Identifier, llvm::VersionTuple,
        StringRef))
+ERROR(serialization_non_ossa_module_incompatible, Fatal,
+      "cannot import non-ossa module into an ossa-module",
```
Comment (atrick) [inline include/swift/AST/DiagnosticsSema.def:905]: Maybe:
"cannot import non-OSSA module into an OSSA module"

**Review by atrick — COMMENTED** [review]
Basically looks good to me

File: include/swift/AST/DiagnosticsSema.def (hunk 5 above)
Comment (meg-gupta) [reply include/swift/AST/DiagnosticsSema.def]: Done

File: include/swift/Serialization/SerializationOptions.h (hunk 3 above)
Comment (meg-gupta) [reply include/swift/Serialization/SerializationOptions.h:27]: Done

File: lib/Frontend/ModuleInterfaceLoader.cpp (hunk 4 above)
Comment (meg-gupta) [reply lib/Frontend/ModuleInterfaceLoader.cpp:1789]: I added a scan-dependencies test in `[enable-ossa-modules-pass-down.swift](https://github.com/swiftlang/swift/pull/77314#diff-ce6adec4434998d49fa1d6b83e2afbd3ecb263c5d57aabb529ab1463576c7cf2)`.
Does that work ?

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 2 above)
Comment (tshortli) [reply lib/Serialization/SerializedModuleLoader.cpp:1195]: Yes, thanks!

Comment (tshortli) [reply lib/Serialization/SerializedModuleLoader.cpp:1195]: Would it be possible to instead introduce a flag that allows OSSA to be lenient for testing, while making both explicit and implicit module builds strict in the same way by default?

**Review by tshortli — APPROVED** [review]

**Review by ktoso — APPROVED** [review]
Not entirely sure on the implications for the Distributed module but seems fine to me?

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/9492

@swift-ci Please test

**Comment by meg-gupta** [comment]
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/9492

@swift-ci Please test

**Comment by meg-gupta** [comment]
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/9492

@swift-ci Please test windows platform

**Comment by felipepiovezan** [comment]
Hi @meg-gupta , this PR broke the LLDB builds: https://ci.swift.org/view/LLDB/job/oss-lldb-incremental-macos-cmake/8289/console

Whenever an API in `ASTContext.h` changes, it is advisable to build LLDB as well, as it is one client of that API.

**Comment by adrian-prantl** [comment]
@shahmishal Why didn't the pre-merge CI catch this?

**Comment by shahmishal** [comment]
@adrian-prantl the llvm change it was tested with was never merged
